File: backend/app/services/providers/factory.py
from __future__ import annotations

from fastapi import HTTPException, status

from app.core.config import settings
from app.models.provider import Provider
from app.services.providers.base import (
    AnthropicAdapter,
    GeminiAdapter,
    OllamaAdapter,
    OpenAIAdapter,
    ProviderAdapter,
)


class AdapterFactory:
    @staticmethod
    def create(provider: Provider, api_key: str) -> ProviderAdapter:
        provider_type = provider.provider_type.lower()

        if provider_type == "openai":
            return OpenAIAdapter(api_key=api_key)
        elif provider_type == "openai_compatible":
            return OpenAIAdapter(api_key=api_key, base_url=provider.base_url)
        elif provider_type == "openrouter":
            return OpenAIAdapter(api_key=api_key, base_url=provider.base_url or "https://openrouter.ai/api/v1")
        elif provider_type == "litellm":
            return OpenAIAdapter(api_key=api_key, base_url=provider.base_url or "http://localhost:4000/v1")
        elif provider_type == "anthropic":
            return AnthropicAdapter(api_key=api_key, base_url=provider.base_url)
        elif provider_type == "gemini":
            return GeminiAdapter(api_key=api_key, base_url=provider.base_url)
        elif provider_type == "ollama":
            base_url = provider.base_url or settings.ollama_base_url
            return OllamaAdapter(base_url=base_url)
        else:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Unsupported provider type: {provider_type}",
            )

File: frontend/core/providers/factory.ts
import { badRequest } from "../errors.js";
import { getConfig } from "../config/index.js";
import { AnthropicAdapter } from "./adapters/anthropic.js";
import { GeminiAdapter } from "./adapters/gemini.js";
import { OllamaAdapter } from "./adapters/ollama.js";
import { OpenAIAdapter } from "./adapters/openai.js";
import type { ProviderAdapter } from "./types.js";

export interface ProviderLike {
  name: string;
  providerType: string;
  baseUrl: string | null;
}

export class AdapterFactory {
  static create(provider: ProviderLike, apiKey: string): ProviderAdapter {
    const providerType = provider.providerType.toLowerCase();
    const providerLabel = provider.name || provider.providerType[0]?.toUpperCase() + provider.providerType.slice(1);

    switch (providerType) {
      case "openai":
        return new OpenAIAdapter({ apiKey, providerLabel });
      case "openrouter":
        return new OpenAIAdapter({ apiKey, baseUrl: provider.baseUrl || "https://openrouter.ai/api/v1", providerLabel });
      case "litellm":
        return new OpenAIAdapter({ apiKey, baseUrl: provider.baseUrl || "http://localhost:4000/v1", providerLabel });
      case "local":
        return new OpenAIAdapter({ apiKey, baseUrl: provider.baseUrl || "http://localhost:4000/v1", providerLabel });
      case "anthropic":
        return new AnthropicAdapter({ apiKey, baseUrl: provider.baseUrl });
      case "gemini":
        return new GeminiAdapter({ apiKey, baseUrl: provider.baseUrl });
      case "ollama":
        return new OllamaAdapter(provider.baseUrl || getConfig().ollamaBaseUrl);
      default:
        throw badRequest(`Unsupported provider type: ${providerType}`);
    }
  }
}
